src/geometry/PolySetBuilder.cc

```diff
@@ -0,0 +1,148 @@
+/*
+ *  OpenSCAD (www.openscad.org)
+ *  Copyright (C) 2009-2011 Clifford Wolf <clifford@clifford.at> and
+ *                          Marius Kintel <marius@kintel.net>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  As a special exception, you have permission to link this program
+ *  with the CGAL library and distribute executables, as long as you
+ *  follow the requirements of the GNU GPL in regard to all of the
+ *  software in the executable aside from CGAL.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ */
+
+#include "PolySetBuilder.h"
+#include <PolySet.h>
+#include "Geometry.h"
+#include "cgalutils.h"
+
+
+#include <memory>
+
+#ifdef ENABLE_MANIFOLD
+#include "ManifoldGeometry.h"
+#endif
+
+
+#include "CGAL_Nef_polyhedron.h"
+#include "CGALHybridPolyhedron.h"
+
+PolySetBuilder::PolySetBuilder(int vertices_count, int indices_count, int dim, boost::tribool convex)
+{
+  ps= new PolySet(dim, convex);
+  if(vertices_count != 0) ps->vertices.reserve(vertices_count);
+  if(indices_count != 0) ps->indices.reserve(indices_count);
+}
+
+PolySetBuilder::PolySetBuilder(const Polygon2d pol)
+{
+  ps= new PolySet(2, unknown);
+  ps->polygon=std::move(pol);   
+  ps->dirty=false;
+}
+
+int PolySetBuilder::vertexIndex(const Vector3d &pt)
+{
+  return allVertices.lookup(pt);
+}
+
+void PolySetBuilder::append_poly(const std::vector<int> &inds)
```

Comment:
I wonder if move semantics could make this more efficient, but we can revisit that later.